### PR TITLE
Add `have_acl_grant()` `its(:acl_grants_count)` `its(:acl_owner)` to s3 type

### PR DESCRIPTION
**Pull request by k1LoW**
see #77

### CODE REVIEW

**Comment by k1LoW** [comment]
AMC UI
screncaptured_2015-11-30_9 34 49

**Comment by k1LoW** [comment]
ACL overview

https://docs.aws.amazon.com/ja_jp/AmazonS3/latest/dev/acl-overview.html

**Comment by k1LoW** [comment]
```ruby
describe s3('my-bucket') do
  it { should exist }
  it { should have_object('path/to/object') }
  it { should have_acl(grantee: 'my-bucket-grantee', permission: 'FULL_CONTROL') }
  its(:bucket_acl_owener) { should eq 'my-bucket-owner' }
end
```

**Comment by kirikiriyamama** [comment]
It's nice!! Then, I have four proposals.

1. I think that `have_acl` should be renamed to `have_bucket_acl_grants` because of the two reasons; (1) a bucket has only one ACL, therefore I don't want to write multiple `have_acl`, and (2) there are two types of ACL's, "bucket" and "object".

2. I'd like to have `have_bucket_acl_grants` receive an Array, for instance `it { have_bucket_acl_grants [{ permission: 'FULL_CONTROL', ... }, ...] }`. If a passed array differs in the number of elements from an API response, a warning will be given (or the test will fail). I want to test either an unexpected ACL exists or not.

3. `type` should be in `grantee` like `{ grantee: { display_name: '', type: '' } }`. I think we should follow the  structure of the API responses.

4. I want to use `its('bucket_acl_owner.display_name')` instead of `its(:bucket_acl_owner)` for the same reason as the 3rd proposal. I know this isn't beautiful. However, we try to test Bucket ACL's by using type S3, so I think we can't avoid the problem.

**Comment by k1LoW** [comment]
Thank you for your proposals!!

> I think that have_acl should be renamed to have_bucket_acl_grants because of the two reasons; (1) a bucket has only one ACL, therefore I don't want to write multiple have_acl, and (2) there are two types of ACL's, "bucket" and "object".

"a bucket has only one ACL" is exactly. I rename `have_acl` to `have_bucket_acl_grant` and `have_grant`. not `have_bucket_acl_grant"s"`.

## Why `have_grant` method_alias

1. `s3` type is bucket resource type. (maybe rename `s3_bucket`)
2. I like clear name.

## Why singular name

use like AMC UI

```
describe s3('my-bucket') do
  it { should have_grant(grantee: 'full-control-user', permission: 'FULL_CONTROL') }
  it { should have_grant(grantee: 'write-only-user', permission: 'WRITE') }
end
```

> I'd like to have have_bucket_acl_grants receive an Array, for instance it { have_bucket_acl_grants [{ permission: 'FULL_CONTROL', ... }, ...] }. If a passed array differs in the number of elements from an API response, a warning will be given (or the test will fail). I want to test either an unexpected ACL exists or not.

I add `bucket_acl_grant_count` like this

```
describe s3('my-bucket') do
  its(:bucket_acl_grant_count) { should eq 5 }
end
```

like `security_group` `its(:inbound_rule_count)` ,  https://github.com/k1LoW/awspec/blob/master/doc/resource_types.md#have_record_set `  its(:resource_record_set_count)`

> type should be in grantee like { grantee: { display_name: '', type: '' } }. I think we should follow the structure of the API responses.

> I want to use its('bucket_acl_owner.display_name') instead of its(:bucket_acl_owner) for the same reason as the 3rd proposal. I know this is not good. However, we try to test Bucket ACL's by using type s3, so I think we can't avoid the problem.

I think awspec should not be affected by API responces. because

1. User do not understand API responces.
2. AWS API are not standardized.

**Comment by kirikiriyamama** [comment]
1. The subject is a bucket as you said. Now, I feel `have_bucket_acl_grant` can be renamed to `have_acl_grant`. Instead, how about creating type `s3_object`? `S3Object#exists?` will be used in place of `S3#has_object?`. If we can remove context of Object from type `s3`, I think it will be simple.

2. I see. I follow existing ways.

3. A `type` does NOT belong to a `grant`, it belongs to a `grantee`. I think that properties of `grantee` should  be nested within `grantee`, right?

4. I agree with you, but I'd like to make the same interface as the 3rd. By the way, `Ec2#iam_instance_profile` returns a `Aws::EC2::Types::IamInstanceProfile`, are you sure to allow this?

**Comment by k1LoW** [comment]
1.

> The subject is a bucket as you said. Now, I feel have_bucket_acl_grant can be renamed to have_acl_grant. 

you're right! I rename.

> how about creating type s3_object?

I think s3_object is good type. Because AWS user can distinguish between the 'bucket' and 'object'.
(And, I want to test object ACL!!)

> S3Object#exists? will be used in place of S3#has_object?

yes. `S3Object#exists?` == `S3#has_object?`. However, it should be present both of which.
awspec use `has_*` and `belong_to_*` to represent "AWS resources relationship".

- `S3Object#exists?` ... existence
- `S3#has_object?`... relationship

3.

> A type does NOT belong to a grant, it belongs to a grantee. I think that properties of grantee should be nested within grantee, right?

That's true. I rename `type` to `grantee_type`. thank you!

4. 

> but I'd like to make the same interface as the 3rd.

I think 

- awspec should have simple usage.
- but awspec may have advanced usage.

> By the way, Ec2#iam_instance_profile returns a Aws::EC2::Types::IamInstanceProfile, are you sure to allow this?

mmm. I had not thought about it until now...

**Comment by kirikiriyamama** [comment]
@k1LoW said, (this is my translation of http://twitter.com/k1LoW/status/671670795689549829)

> If we'd like to test all of obtainable information from API, we should use aws-sdk on RSpec. Rather, we should test only important points in AWS resources.

I respect the idea, so I rethought my opinions. Now, I agree with this interface https://github.com/k1LoW/awspec/pull/79#issuecomment-160531501 . I'm sorry for the backtracking.

**Comment by kirikiriyamama** [comment]
Let me know if my translation is wrong.

**Comment by k1LoW** [comment]
@kirikiriyamama Thank you for your advice!

Finally, like this

```ruby
describe s3('my-bucket') do
   it { should exist }
   it { should have_object('path/to/object') }
   its(:acl_grant_count) { should eq 3 }
   it { should have_acl_grant(grantee: 'my-bucket-owner', permission: 'FULL_CONTROL', grantee_type: 'CanonicalUser') }
   it { should have_acl_grant(grantee: 'my-bucket-write-only', permission: 'WRITE', grantee_type: 'CanonicalUser') }
   it { should have_acl_grant(grantee: 'my-bucket-read-only', permission: 'READ', grantee_type: 'CanonicalUser') }
   its(:acl_owner) { should eq 'my-bucket-owner' }
 end
```
( remove `bucket_` prefix. rename `type` to `grantee_type`. in favor of https://github.com/k1LoW/awspec/pull/79#issuecomment-160938516 1,3 )

I will merge this PR !!

**Comment by kirikiriyamama** [comment]
Ah... I think important points of a `grant`, are `grantee.display_name` and `permission`. Therefore `grantee_type` can be removed. What about you? I'm sorry for the backtracking.

File: lib/awspec/type/s3.rb [hunk 1]
```diff
@@ -15,5 +15,25 @@ def has_object?(key)
...
+        true
+      end
+    end
+
+    def acl_owner
+      @acl = find_s3_bucket_acl(@id)
+      @acl.owner.display_name
+    end
+
+    def acl_grant_count
```
Comment (kirikiriyamama) [inline lib/awspec/type/s3.rb]: I think `acl_grants_count` is better than `acl_grant_count`.

**Comment by k1LoW** [comment]
hum. you're right.

File: lib/awspec/helper/finder/s3.rb [hunk 2]
```diff
@@ -7,6 +7,12 @@ def find_bucket(id)
           bucket.name == id
         end
       end
+
+      def find_s3_bucket_acl(id)
```
Comment (kirikiriyamama) [inline lib/awspec/helper/finder/s3.rb]: It's clear that a Bucket ACL belongs to S3. How about `find_bucket_acl`?

**Comment by k1LoW** [comment]
When user became want to test `grant.type`, I will add `grantee_type`.

File: lib/awspec/helper/finder/s3.rb (hunk 2 above)
Comment (k1LoW) [inline lib/awspec/helper/finder/s3.rb]: oh..

Comment (k1LoW) [inline lib/awspec/helper/finder/s3.rb]: renamed!

File: lib/awspec/type/s3.rb (hunk 1 above)
Comment (k1LoW) [inline lib/awspec/type/s3.rb]: oh..oh..
that's true.

**Comment by kirikiriyamama** [comment]
If anyone want to test `grantee.type`, I think https://github.com/k1LoW/awspec/pull/81 should be used, shouldn't it?

**Comment by k1LoW** [comment]
> If anyone want to test grantee.type, I think #81 should be used, shouldn't it?

!!!! you are right !!!! :bow:

**Comment by k1LoW** [comment]
once again

```ruby
describe s3('my-bucket') do
   it { should exist }
   it { should have_object('path/to/object') }
   its(:acl_grant_count) { should eq 3 }
   it { should have_acl_grant(grantee: 'my-bucket-owner', permission: 'FULL_CONTROL') }
   it { should have_acl_grant(grantee: 'my-bucket-write-only', permission: 'WRITE') }
   it { should have_acl_grant(grantee: 'my-bucket-read-only', permission: 'READ') }
   its(:acl_owner) { should eq 'my-bucket-owner' }
 end
```

**Comment by kirikiriyamama** [comment]
It's awesome!!! Thank you for everything :sparkles:

**Comment by kirikiriyamama** [comment]
`s/acl_grant_count/acl_grants_count/` ?

**Comment by k1LoW** [comment]
Oops! comment miss.

The following code is "the truth".
```
describe s3('my-bucket') do
   it { should exist }
   it { should have_object('path/to/object') }
   its(:acl_grants_count) { should eq 3 }
   it { should have_acl_grant(grantee: 'my-bucket-owner', permission: 'FULL_CONTROL') }
   it { should have_acl_grant(grantee: 'my-bucket-write-only', permission: 'WRITE') }
   it { should have_acl_grant(grantee: 'my-bucket-read-only', permission: 'READ') }
   its(:acl_owner) { should eq 'my-bucket-owner' }
 end
```

merge!!

**Comment by k1LoW** [comment]
Released as v0.23.0.

**Pull request merged**